baseapp/manager.go

```diff
@@ -2,113 +2,164 @@ package baseapp
+	m.executorCfg.PrometheusPrefix = executorPromName
+
+	return m
+}
+
+// Logger returns the logger for the baseapp.
+func (jm *Manager) Logger(ctx context.Context) log.Logger {
+	return sdk.UnwrapSdkContext(ctx).Logger().With("namespace", "job-manager")
+}
+
+// Start calls `Setup` on the jobs in the registry as well as spins up
+// the worker pools.
+func (jm *Manager) Start(ctx context.Context) {
+	// We pass in the context in order to handle cancelling the workers.
+	jm.jobExecutors = worker.NewPool(ctx, jm.executorCfg)
+	jm.jobProducers = worker.NewPool(ctx, jm.producerCfg)
+
+	for _, j := range jm.jobRegistry.Iterate() {
+		if sj, ok := j.(job.HasSetup); ok {
+			if err := sj.Setup(ctx); err != nil {
+				panic(err)
+			}
+		}
+	}
+}
+
+// Stop calls `Teardown` on the jobs in the registry as well as
+// shut's down all the worker pools.
+func (jm *Manager) Stop() {
+	for _, j := range jm.jobRegistry.Iterate() {
+		if tj, ok := j.(job.HasTeardown); ok {
+			if err := tj.Teardown(); err != nil {
+				panic(err)
+			}
+		}
+	}
+
+	jm.jobProducers.Stop()
+	jm.jobExecutors.Stop()
+	jm.jobProducers = nil
+	jm.jobExecutors = nil
+}
+
+// RunProducers runs the job producers.
+//
+//nolint:gocognit // fix.
+func (jm *Manager) RunProducers(ctx context.Context) {
+	for _, j := range jm.jobRegistry.Iterate() {
+		// Handle migrated jobs.
+		if wrappedJob := job.WrapJob(j); wrappedJob != nil {
+			jm.jobProducers.Submit(
+				func() {
+					if err := wrappedJob.Producer(ctx, jm.jobExecutors); err != nil {
+						jm.Logger(ctx).Error("error in job producer", "err", err)
 					}
-				}
-			}()
-		} else if subJob, ok := j.(job.Subscribable); ok { //nolint:govet // todo fix.
-			go func() {
+				},
+			)
+			continue
+		}
+
+		// Handle unmigrated jobs. // TODO: migrate format.
+		if subJob, ok := j.(job.Subscribable); ok {
+			jm.jobExecutors.Submit(func() {
 				ch := subJob.Subscribe(ctx)
 				for {
 					select {
 					case val := <-ch:
-						jm.executionPool.AddTask(job.NewExecutor(ctx, subJob, val))
+						jm.jobExecutors.Submit(workertypes.NewPayload(ctx, subJob, val).Execute)
 					case <-ctx.Done():
 						return
 					default:
 						continue
 					}
 				}
-			}()
+			})
+			// Handle unmigrated jobs. // TODO: migrate format.
 		} else if ethSubJob, ok := j.(job.EthSubscribable); ok { //nolint:govet // todo fix.
-			go func() {
+			jm.jobExecutors.Submit(func() {
 				sub, ch := ethSubJob.Subscribe(ctx)
 				for {
 					select {
 					case <-ctx.Done():
 						ethSubJob.Unsubscribe(ctx)
 						return
 					case err := <-sub.Err():
-						jm.logger.Error("error in subscription", "err", err)
+						jm.Logger(ctx).Error("error in subscription", "err", err)
 						// TODO: add retry mechanism
 						ethSubJob.Unsubscribe(ctx)
 						return
 					case val := <-ch:
-						jm.executionPool.AddTask(job.NewExecutor(ctx, ethSubJob, val))
+						jm.jobExecutors.Submit(workertypes.NewPayload(ctx, ethSubJob, val).Execute)
 						continue
 					}
 				}
```

Comment:
calling to `Subscribe` and `EthSubscribe` should have diff underlying contexts.

Reply:
going to handle in a later PR tbh.